### PR TITLE
Ep bonus command

### PR DESCRIPTION
**Pull request by MikeJeffers**
Closes: https://github.com/OpenPerpetuum/OP-Project/issues/114

REQUIRES: https://github.com/OpenPerpetuum/OPDB/pull/92

### CODE REVIEW

**Review by LoyalServant — DISMISSED** [review]

**Comment by MikeJeffers** [comment]
Alright guys, the first commit was no good because I ignored concurrency issues.
There are several ways to fix this, which will have some effect on performance, safety/robustness, maintainability, and style.

I chose to go with the Perpetuum Style Process.Update handling, because the use case that we are using is quite similar.

The use case: event(requestHandler)->start server "event" (a temporal setting on the server that will expire)
In update loop, track elapsing time at some frequency (set in bootstrapper) when the elapsed time exceeds the intended duration, "end" the event.

[TimerAction](https://github.com/OpenPerpetuum/PerpetuumServer/blob/Development/src/Perpetuum/Timers/TimerAction.cs) achieves this, but we need to create and destroy these objects, and check for null on the update loop.  This requires a mutex or lock.

This latest commit also follows the PerpStyle on these update loops that track time-based events.
Take a look at: [RoamingPresence](https://github.com/OpenPerpetuum/PerpetuumServer/blob/Development/src/Perpetuum/Zones/NpcSystem/Presences/RoamingPresence.cs#L67)

Also bonus, no need to create and destroy TimerAction objects or keep track of a lock object instance. (Though this is a singleInstance service).  

I graciously await your review comrades.

File: src/Perpetuum/Services/EventServices/EPBonusEventService.cs
Comment (nref-dan) [inline src/Perpetuum/Services/EventServices/EPBonusEventService.cs]: I would set this as readonly

File: src/Perpetuum.RequestHandlers/EPBonusEvent.cs
Comment (nref-dan) [inline src/Perpetuum.RequestHandlers/EPBonusEvent.cs]: We should declare these as constants

File: src/Perpetuum.RequestHandlers/EPBonusEvent.cs
Comment (nref-dan) [inline src/Perpetuum.RequestHandlers/EPBonusEvent.cs:27]: How about also storing the date we want the event to start from (in international yyyy-MM-dd format)? That way an admin doesn't need to be online at X time to begin the event.

File: src/Perpetuum/Accounting/AccountManager.cs
Comment (nref-dan) [inline src/Perpetuum/Accounting/AccountManager.cs]: In an ideal world the boost factor should be retrieved from the service providing the boost event.

File: src/Perpetuum/Accounting/AccountManager.cs
Comment (nref-dan) [inline src/Perpetuum/Accounting/AccountManager.cs]: This is the right place to add the bonus, but it makes the logging incorrect. On line 366 the boost factor is calculated using the old linear scaling algorithm, but this doesn't take into account the bonus that's just been added.

File: src/Perpetuum/Commands.cs
Comment (nref-dan) [inline src/Perpetuum/Commands.cs:3477]: Not anything wrong with this but just a note that you can do the following:

`public static Command EPBonusSet => new Command { ... };`

Which has the added bonus of it now being a read only property, newly instantiated only when needed and destroyed once finished. If CPU cycles are a concern then obviously this isn't a good idea :)

File: src/Perpetuum/Services/EventServices/EPBonusEventService.cs
Comment (nref-dan) [inline src/Perpetuum/Services/EventServices/EPBonusEventService.cs]: Try and avoid using this pattern; it can behave unexpectedly. For example in MVC, the ContinueWith would never execute because there is a change of SynchronisationContext between the executor and the target. Or in shared TargetSynchronisationContexts (typically WinForms apps) you can deadlock as it waits to join the thread that's waiting for it to join.

It seems like you want Update to finish executing (so as to not hang up the Server event loop) but you want DoClearBonus and the variables to be executed synchronously.

I would in this case, simply declare the entire block within the Run statement, or, create a method and call that.

eg. 
```
Task.Run(() =>
{
    DoClearBonus();
    _eventStart = false;
    _endingEvent = false;
    _elapsed = TimeSpan.Zero;
});
```
If you don't care about the synchronous execution of `DoClearBonus()` then you can make the Runner asynchronous.
```
Task.Run(async () =>
{
    await Task.Run(() => DoClearBonus());
    _eventStart = false;
    _endingEvent = false;
    _elapsed = TimeSpan.Zero;
});
```

File: src/Perpetuum/Accounting/AccountManager.cs
Comment (MikeJeffers) [reply src/Perpetuum/Accounting/AccountManager.cs]: Maybe I am misunderstanding, but boostingfactor tracks only along the curve of ep-on-account over the target, (the inverse of that) clamped to [0.-1.].  The multiplier is applied after, while assumed to be constant, could be anything with the new event.  But it can still be computed from the logging w/ the raw and boosted points.
To me they are separate, and didn't think to alter the boostingFactor to compensate for an EP Bonus Event, which to keep it simple, pushes the Multiplier up.

Is there a dependency on the epForActivityLog that I am missing?

Comment (nref-dan) [reply src/Perpetuum/Accounting/AccountManager.cs]: image

Not exactly. Line 367 calls GetBoostedExtensionPoints, this in turn calls GetEPCollected and AddEPWithBoosting. AddEPWithBoosting gets the EP Boosting factor and then re-calculated this with the multiplier to give you an overall EP gain. This is returned as the "real EP gain" and is populated into the boostedPoints variable. Now the boost factor contains the *non* boost multiplied factor (the old scaling number) but the boosted points value contains the multiplied EP. If you were to try and back calculated the EP from the log, this number would be inconsistent with what's present in the log. When the LogEpForActivity call audits this into the database, the rawPoints and actualPoints will match what were given, however, the boost factor will not. Whilst you can back-calculate this, it's not evident at a glance why the numbers wouldn't match (think new non-technical support agent)

So in order to keep the numbers consistent either;
a) We need to record an extra value in the log indicating the boost multiplication value
b) We need to pass in the correct boost factor to the audit log

So just to be clear, I'm not saying there's a functionality problem with the boost factor and how it's applied, however, the audit trail will be inconsistent and/or confusing because this information isn't being recorded.

**Review by nref-dan — COMMENTED** [review]
Styling Comment: If you only specify a method in the Task.Run lambda you don't need the curly braces

File: src/Perpetuum/Services/EventServices/EPBonusEventService.cs
Comment (nref-dan) [inline src/Perpetuum/Services/EventServices/EPBonusEventService.cs]: Styling Comment: If you only specify a method in the Task.Run lambda you don't need the curly braces

File: src/Perpetuum/Accounting/AccountManager.cs
Comment (MikeJeffers) [reply src/Perpetuum/Accounting/AccountManager.cs]: I see if we wanted to avoid adding a new column we would need to make the boostfactor an "effective boostfactor" so a 25x during a +5 event shows 1.20 boostfactor. (?)
That way any existing methods or recovery procs on this log wont need to be tracked down and updated.  
Would you say A) then is the better option here?

**Review by nref-dan — CHANGES_REQUESTED** [review]
If we're going to lock for concurrency then I suggest using the `ReaderWriterLockSlim` class ( https://docs.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlockslim?view=netframework-4.7.2 )

With a standard `lock` your read contention will be single threaded, however, we only care about write contention. In this system there will be a ratio of probably 1000's to 1 for reads vs writes. As such, we don't want to lock threads from reading *unless* a write is occurring. The ReaderWriterLockSlim class will allow us to lock the reads behind a read lock and the write behind a write lock, allowing multi-threaded concurrent reading, but single threaded write access, blocking the reads until the write has finished.

**Comment by MikeJeffers** [comment]
See most recent commit for use of the ReaderWriterLockSlim
https://github.com/OpenPerpetuum/PerpetuumServer/pull/77/commits/f1af82f13914450e2dc12184b43fab3553aac22a

File: src/Perpetuum/Services/EventServices/EPBonusEventService.cs
Comment (nref-dan) [inline src/Perpetuum/Services/EventServices/EPBonusEventService.cs]: `_eventStarted` and `_endingEvent` need protecting by a read lock, or by moving the write lock up to line 113. Personally I'd go with a read lock.

**Review by nref-dan — CHANGES_REQUESTED** [review]
Just one last change and then I'll sign off my review

**Review by nref-dan — DISMISSED** [review]

**Comment by nref-dan** [comment]
My changes need reviewing before this is merged

**Comment by MikeJeffers** [comment]
Bug on the new threadlock stuff

```
[20:12:30] ERR
Exception Found:
Type: System.ObjectDisposedException
Message: Cannot access a disposed object.
Source: System.Core
Stacktrace:    at System.Threading.ReaderWriterLockSlim.TryEnterReadLockCore(TimeoutTracker timeout)
   at System.Threading.ReaderWriterLockSlim.TryEnterReadLock(TimeoutTracker timeout)
   at System.Threading.ReaderWriterLockSlim.TryEnterReadLock(TimeSpan timeout)
   at Perpetuum.Threading.ThreadLockExtensions.WriteLock..ctor(ReaderWriterLockSlim syncLock, TimeSpan timeout) in C:\Users\Unavailable\Documents\OpenPerpetuum\PerpetuumServer\src\Perpetuum\Threading\ThreadLock.cs:line 45
   at Perpetuum.Threading.ThreadLockExtensions.Write(ReaderWriterLockSlim obj, TimeSpan timeout) in C:\Users\Unavailable\Documents\OpenPerpetuum\PerpetuumServer\src\Perpetuum\Threading\ThreadLock.cs:line 70
   at Perpetuum.Services.EventServices.EPBonusEventService.Init() in C:\Users\Unavailable\Documents\OpenPerpetuum\PerpetuumServer\src\Perpetuum\Services\EventServices\EPBonusEventService.cs:line 29
   at Perpetuum.Services.EventServices.EPBonusEventService.Start() in C:\Users\Unavailable\Documents\OpenPerpetuum\PerpetuumServer\src\Perpetuum\Services\EventServices\EPBonusEventService.cs:line 99
   at Perpetuum.Threading.Process.ProcessDecorator.Start() in C:\Users\Unavailable\Documents\OpenPerpetuum\PerpetuumServer\src\Perpetuum\Threading\Process\TimedProcess.cs:line 19
   at Perpetuum.Threading.Process.ProcessDecorator.Start() in C:\Users\Unavailable\Documents\OpenPerpetuum\PerpetuumServer\src\Perpetuum\Threading\Process\TimedProcess.cs:line 19
   at Perpetuum.Threading.Process.ProcessManager.Start() in C:\Users\Unavailable\Documents\OpenPerpetuum\PerpetuumServer\src\Perpetuum\Threading\Process\ProcessManager.cs:line 46
   at Perpetuum.Bootstrapper.PerpetuumBootstrapper.<Init>b__9_1(IHostStateService sender, HostState state) in C:\Users\Unavailable\Documents\OpenPerpetuum\PerpetuumServer\src\Perpetuum.Bootstrapper\PerpetuumBootstrapper.cs:line 348
   at System.Action`2.Invoke(T1 arg1, T2 arg2)
   at Perpetuum.Host.HostStateService.OnStateChanged(HostState state) in C:\Users\Unavailable\Documents\OpenPerpetuum\PerpetuumServer\src\Perpetuum\Host\HostStateService.cs:line 33
```
Initial trace shows on 2nd call into Init the _lock object's fDisposed property is true.

**Comment by MikeJeffers** [comment]
Tested and working!
(Please confirm)

**Review by LoyalServant — APPROVED** [review]

**Pull request merged**
